File: extra_1/app/src/main/java/com/example/extra_1/ui/WeatherViewModel.java
package com.example.extra_1.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.extra_1.api.NetworkSingleton;
import com.example.extra_1.model.WeatherModels;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class WeatherViewModel extends ViewModel {
    private final MutableLiveData<String> today = new MutableLiveData<>();
    private final MutableLiveData<List<WeatherModels.DayForecast>> week = new MutableLiveData<>();

    public LiveData<String> getToday() { return today; }
    public LiveData<List<WeatherModels.DayForecast>> getWeek() { return week; }

    public void load(String city, String apiKey) {
        NetworkSingleton.getWeatherApi().getForecast(city, apiKey, "metric", "ru")
                .enqueue(new Callback<>() {
                    @Override
                    public void onResponse(Call<WeatherModels.ForecastResponse> call, Response<WeatherModels.ForecastResponse> response) {
                        if (!response.isSuccessful()) {
                            if (response.code() == 401) {
                                today.setValue("Не удалось загрузить погоду. Проверьте API-ключ.");
                            } else {
                                today.setValue("Ошибка сервера: " + response.code());
                            }
                            week.setValue(new ArrayList<>());
                            return;
                        }

                        List<WeatherModels.ForecastItem> data = response.body() != null ? response.body().list : new ArrayList<>();
                        if (data == null || data.isEmpty()) {
                            today.setValue("Нет данных о погоде.");
                            week.setValue(new ArrayList<>());
                            return;
                        }

                        WeatherModels.ForecastItem first = data.get(0);
                        String desc = (first.weather != null && !first.weather.isEmpty()) ? first.weather.get(0).description : "";
                        today.setValue(first.main.temp + "°C, " + desc);

                        Map<String, List<Double>> byDay = new LinkedHashMap<>();
                        for (WeatherModels.ForecastItem item : data) {
                            String day = item.dt_txt.split(" ")[0];
                            byDay.computeIfAbsent(day, k -> new ArrayList<>()).add(item.main.temp);
                        }

                        List<WeatherModels.DayForecast> out = new ArrayList<>();
                        int i = 0;
                        for (Map.Entry<String, List<Double>> e : byDay.entrySet()) {
                            if (i++ >= 7) break;
                            double sum = 0;
                            for (Double t : e.getValue()) sum += t;
                            out.add(new WeatherModels.DayForecast(e.getKey(), ((int) (sum / e.getValue().size())) + "°C"));
                        }
                        week.setValue(out);
                    }

                    @Override
                    public void onFailure(Call<WeatherModels.ForecastResponse> call, Throwable t) {
                        today.setValue("Ошибка сети: " + t.getMessage());
                        week.setValue(new ArrayList<>());
                    }
                });
    }
}

File: extra1/app/src/main/java/com/example/extra1/ui/WeatherViewModel.kt
package com.example.extra1.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.extra1.api.NetworkSingleton
import com.example.extra1.model.DayForecast
import com.example.extra1.model.ForecastResponse
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class WeatherViewModel : ViewModel() {
    private val _today = MutableLiveData<String>()
    val today: LiveData<String> = _today
    private val _week = MutableLiveData<List<DayForecast>>()
    val week: LiveData<List<DayForecast>> = _week

    fun load(city: String, apiKey: String) {
        NetworkSingleton.weatherApi.getForecast(city, apiKey).enqueue(object : Callback<ForecastResponse> {
            override fun onResponse(call: Call<ForecastResponse>, response: Response<ForecastResponse>) {
                val data = response.body()?.list.orEmpty()
                if (data.isEmpty()) return
                val todayItem = data.first()
                _today.value = "${todayItem.main.temp}°C, ${todayItem.weather.firstOrNull()?.description ?: ""}"

                val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
                _week.value = data.groupBy { LocalDateTime.parse(it.dt_txt, formatter).toLocalDate().toString() }
                    .entries.take(7).map { (d, items) ->
                        DayForecast(d, "${items.map { it.main.temp }.average().toInt()}°C")
                    }
            }

            override fun onFailure(call: Call<ForecastResponse>, t: Throwable) {
                _today.value = "Ошибка: ${t.message}"
                _week.value = emptyList()
            }
        })
    }
}
